File: Base/Desktop/Tests/WorkspaceTests.ts
namespace Tests {

    export class WorkspaceTests extends tsUnit.TestClass {

        load() {
            const workspace = new Allors.Workspace(Allors.Data.metaPopulation);
            workspace.load(Fixture.loadData);

            var martien = workspace.get("3");

            this.areIdentical("3", martien.id);
            this.areIdentical("1003", martien.version);
            this.areIdentical("Person", martien.objectType.name);
            this.areIdentical("Martien", martien.roles.FirstName);
            this.areIdentical("van", martien.roles.MiddleName);
            this.areIdentical("Knippenberg", martien.roles.LastName);
            this.areIdentical(undefined, martien.roles.IsStudent);
            this.areIdentical(undefined, martien.roles.BirthDate);
        }

        checkVersions() {
            const workspace = new Allors.Workspace(Allors.Data.metaPopulation);
            workspace.userSecurityHash = "#";
            workspace.load(Fixture.loadData);

            var required =
                {
                    userSecurityHash: "#",
                    objects: [
                        ["1", "1001"],
                        ["2", "1002"],
                        ["3", "1004"]
                    ]
                }

            var requireLoad = workspace.diff(required);

            this.areIdentical(1, requireLoad.objects.length);
        }

        checkVersionsUserSecurityHash() {
            const workspace = new Allors.Workspace(Allors.Data.metaPopulation);
            workspace.userSecurityHash = "abc";
            workspace.load(Fixture.loadData);

            var required =
                {
                    userSecurityHash: "def",
                    objects: [
                        ["1", "1001"],
                        ["2", "1002"],
                        ["3", "1004"]
                    ]
                }

            var requireLoad = workspace.diff(required);

            this.areIdentical(3, requireLoad.objects.length);
        }
    }
}

File: Base/Workspace.Domain.Tests/Workspace/WorkspaceTests.cs
namespace Tests
{
    using Allors.Data;
    using Allors.Workspace;
    using NUnit.Framework;

    [TestFixture]
    public class WorkspaceTests
    {
        [Test]
        public void Load()
        {
            var workspace = new Workspace(Config.ObjectFactory);
            workspace.sync(Fixture.loadData);

            var martien = workspace.get("3");

            Assert.AreEqual("3", martien.id);
            Assert.AreEqual("1003", martien.version);
            Assert.AreEqual("Person", martien.objectType.Name);
            Assert.AreEqual("Martien", martien.roles["FirstName"]);
            Assert.AreEqual("van", martien.roles["MiddleName"]);
            Assert.AreEqual("Knippenberg", martien.roles["LastName"]);
            Assert.IsFalse(martien.roles.ContainsKey("IsStudent"));
            Assert.IsFalse(martien.roles.ContainsKey("BirthDate"));
        }

        [Test]
        public void CheckVersions()
        {
            var workspace = new Workspace(Config.ObjectFactory)
                                {
                                    userSecurityHash = "#"
                                };
            workspace.sync(Fixture.loadData);

            var required = new PullResponse
                               {
                                   userSecurityHash = "#",
                                   objects =
                                       new[]
                                           {
                                                new[] { "1", "1001" },
                                                new[] { "2", "1002" },
                                                new[] { "3", "1004" }
                                           }
                               };

            var requireLoad = workspace.diff(required);

            Assert.AreEqual(1, requireLoad.objects.Length);
        }

        [Test]
        public void CheckVersionsUserSecurityHash()
        {
            var workspace = new Workspace(Config.ObjectFactory) { userSecurityHash = "abc" };
            workspace.sync(Fixture.loadData);

            var required = new PullResponse
                               {
                                   userSecurityHash = "def",
                                   objects =
                                       new[]
                                           {
                                                new[] { "1", "1001" },
                                                new[] { "2", "1002" },
                                                new[] { "3", "1004" }
                                           }
                               };

            var requireLoad = workspace.diff(required);

            Assert.AreEqual(3, requireLoad.objects.Length);
        }
    }
}